packages/headless/src/features/result-templates/result-templates-manager.ts

```diff
@@ -1,72 +1,99 @@
- */
-export class ResultTemplatesManager<Content = unknown, State = SearchAppState> {
-  private templates: Required<ResultTemplate<Content>>[] = [];
-  /**
-   * Creates a new `ResultTemplatesManager` instance.
-   * @param engine (HeadlessEngine) The `HeadlessEngine` instance of your application.
-   * @returns (ResultTemplatesManager<Content, State>) A new result templates manager.
-   */
-  constructor(private engine: Engine<State>) {}
-
+export interface ResultTemplatesManager<Content = unknown> {
   /**
    * Registers any number of result templates in the manager.
    * @param templates (...ResultTemplate<Content>) A list of templates to register.
    */
-  public registerTemplates(...templates: ResultTemplate<Content>[]) {
-    const fields: string[] = [];
-    this.validateTemplates(templates);
-
-    templates.forEach((template) => {
-      const templatesWithDefault = {
-        ...template,
-        priority: template.priority || 0,
-        fields: template.fields || [],
-      };
-      this.templates.push(templatesWithDefault);
-      fields.push(...templatesWithDefault.fields);
-    });
-
-    this.templates.sort((a, b) => b.priority - a.priority);
-
-    fields.length && this.engine.dispatch(registerFieldsToInclude(fields));
-  }
+  registerTemplates(...newTemplates: ResultTemplate<Content>[]): void;
+  /**
+   * Selects the highest priority template for which the given result satisfies all conditions.
+   * In the case where satisfied templates have equal priority, the template that was registered first is returned.
+   * @param result (Result) The result for which to select a template.
+   * @returns (Content) The selected template's content, or null if no template's conditions are satisfied.
+   */
+  selectTemplate(result: Result): Content | null;
+}
 
-  private validateTemplates(templates: ResultTemplate<Content>[]) {
+/**
+ * A manager in which result templates can be registered and selected based on a list of conditions and priority.
+ * @param engine (HeadlessEngine) The `HeadlessEngine` instance of your application.
+ * @returns (ResultTemplatesManager<Content, State>) A new result templates manager.
+ */
+export function buildResultTemplatesManager<
+  Content = unknown,
+  State = SearchAppState
+>(engine: Engine<State>): ResultTemplatesManager<Content> {
+  const templates: Required<ResultTemplate<Content>>[] = [];
+  const validateTemplates = (templates: ResultTemplate<Content>[]) => {
     templates.forEach((template) => {
-      prioritySchema.validate(template);
+      prioritySchema.validate(
+        template,
+        'Check the arguments of registerTemplates'
+      );
       const areConditionsValid = template.conditions.every(
         (condition) => condition instanceof Function
       );
 
       if (!areConditionsValid) {
-        throw new Error('Result template conditions are invalid');
+        throw new SchemaValidationError('Conditions are invalid');
```

Comment:
The error message here is a bit dry, and does not help very much to solve the problem: "why are they invalid ?", "what should be the format ?", etc.